File: src/FSharpVSPowerTools.Pure/Commands/RecordStubGeneratorSmartTaggerProvider.fs
module FSharpVSPowerTools.RecordStubGeneratorSmartTaggerProvider

open System
open System.ComponentModel.Composition
open Microsoft.VisualStudio.Text
open Microsoft.VisualStudio.Text.Editor
open Microsoft.VisualStudio.Text.Operations
open Microsoft.VisualStudio.Text.Tagging
open Microsoft.VisualStudio.Utilities
open Microsoft.VisualStudio.Shell
open FSharpVSPowerTools.Refactoring
open FSharpVSPowerTools.ProjectSystem
open System.Diagnostics
open Microsoft.VisualStudio.Shell.Interop

[<Export(typeof<IViewTaggerProvider>)>]
[<ContentType "F#">]
[<TagType(typeof<RecordStubGeneratorSmartTag>)>]
type RecordStubGeneratorSmartTaggerProvider [<ImportingConstructor>]
    ( [<Import(typeof<SVsServiceProvider>)>] 
        serviceProvider             :   IServiceProvider            ,
        textDocumentFactoryService  :   ITextDocumentFactoryService ,
        textEditorFactoryService    :   ITextEditorFactoryService   ,
        undoHistoryRegistry         :   ITextUndoHistoryRegistry    ,
        projectFactory              :   ProjectFactory              ,
        vsLanguageService           :   VSLanguageService           )  =

    interface IViewTaggerProvider with
        member __.CreateTagger (textView, buffer) =
            maybe {
                // Only provide the smart tagger on the top-level buffer
                if textView.TextBuffer <> buffer then return! None else
       //         let generalOptions = Setting.getGeneralOptions serviceProvider
         //       if not generalOptions.GenerateRecordStubEnabled then return! None else
                let codeGenOptions = Setting.getCodeGenerationOptions serviceProvider
                let dte = serviceProvider.GetService<EnvDTE.DTE,SDTE>()
                if dte.Version = string VisualStudioVersion.VS2015 then return! None else
                let! doc = textDocumentFactoryService.TryDocumentFromBuffer buffer
                let generator = 
                    new RecordStubGenerator (doc, textView, undoHistoryRegistry.RegisterHistory buffer, 
                        vsLanguageService, serviceProvider, projectFactory, 
                        Setting.getDefaultMemberBody codeGenOptions, OpenDocumentsTracker textDocumentFactoryService)
                return
                    new RecordStubGeneratorSmartTagger (buffer, generator) :> obj :?> _
            } |> Option.getOrElse null

File: src/FSharpVSPowerTools/Commands/RecordStubGeneratorSmartTaggerProvider.cs
﻿using System;
using System.ComponentModel.Composition;
using Microsoft.VisualStudio.Text;
using Microsoft.VisualStudio.Text.Editor;
using Microsoft.VisualStudio.Text.Operations;
using Microsoft.VisualStudio.Text.Tagging;
using Microsoft.VisualStudio.Utilities;
using Microsoft.VisualStudio.Shell;
using FSharpVSPowerTools.Refactoring;
using FSharpVSPowerTools.ProjectSystem;
using Microsoft.VisualStudio.Shell.Interop;

namespace FSharpVSPowerTools
{
    [Export(typeof(IViewTaggerProvider))]
    [ContentType("F#")]
    [TagType(typeof(RecordStubGeneratorSmartTag))]
    [TextViewRole(PredefinedTextViewRoles.Editable)]
    [TextViewRole(PredefinedTextViewRoles.PrimaryDocument)]
    public class RecordStubGeneratorSmartTaggerProvider : IViewTaggerProvider
    {
        private readonly IServiceProvider _serviceProvider;
        private readonly ITextDocumentFactoryService _textDocumentFactoryService;
        private readonly ProjectFactory _projectFactory;
        private readonly VSLanguageService _fsharpVsLanguageService;
        private readonly IOpenDocumentsTracker _openDocumentsTracker;
        private readonly ITextUndoHistoryRegistry _undoHistoryRegistry;
        private readonly IGeneralOptions _generalOptions;
        private readonly ICodeGenerationOptions _codeGenOptions;


        [ImportingConstructor]
        public RecordStubGeneratorSmartTaggerProvider(
            [Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider,
            ITextDocumentFactoryService textDocumentFactoryService,
            ITextUndoHistoryRegistry undoHistoryRegistry,
            ProjectFactory projectFactory,
            VSLanguageService fsharpVsLanguageService,
            IOpenDocumentsTracker openDocumentsTracker,
            IGeneralOptions generalOptions  ,
            ICodeGenerationOptions codeGenOptions)
        {
            _serviceProvider = serviceProvider;
            _textDocumentFactoryService = textDocumentFactoryService;
            _undoHistoryRegistry = undoHistoryRegistry;
            _projectFactory = projectFactory;
            _fsharpVsLanguageService = fsharpVsLanguageService;
            _openDocumentsTracker = openDocumentsTracker;
            _generalOptions = generalOptions;
            _codeGenOptions = codeGenOptions;
        }

        public ITagger<T> CreateTagger<T>(ITextView textView, ITextBuffer buffer) where T : ITag
        {
            // Only provide the smart tagger on the top-level buffer
            if (textView.TextBuffer != buffer) return null;

            //var generalOptions = Setting.getGeneralOptions(_serviceProvider);
            if (_generalOptions == null || !_generalOptions.GenerateRecordStubEnabled) return null;
            //var codeGenOptions = Setting.getCodeGenerationOptions(_serviceProvider);
            if (_codeGenOptions == null) return null;

            var dte = _serviceProvider.GetService(typeof(SDTE)) as EnvDTE.DTE;
            var vsVersion = VisualStudioVersionModule.fromDTEVersion(dte.Version);
            if (vsVersion == VisualStudioVersion.VS2015) return null;

            ITextDocument doc;
            if (_textDocumentFactoryService.TryGetTextDocument(buffer, out doc))
            {
                var generator = new RecordStubGenerator(doc, textView,
                                _undoHistoryRegistry.RegisterHistory(buffer),
                                _fsharpVsLanguageService, _serviceProvider,
                                _projectFactory, _codeGenOptions.DefaultBody,
                                _openDocumentsTracker);
                return new RecordStubGeneratorSmartTagger(buffer, generator) as ITagger<T>;
            }
            
            return null;
        }
    }
}
